### ISSUE TITLE
ACCURACY IS USING NUMPY, URGENT FIX

### ISSUE THREAD

**Issue by zackchase** [apache/incubator-mxnet#9571]
Accuracy is the bottleneck for even a 4000 node neural network. This is a practical job that many people will run and which a GPU ***CAN accelerate***.

But right now the accuracy metric is the bottleneck. That's because it converts everything to numpy, causing a blocking operation at every iteration. Let's fix it stat.

  
```
for label, pred_label in zip(labels, preds):
            if pred_label.shape != label.shape:
                pred_label = ndarray.argmax(pred_label, axis=self.axis)
            pred_label = pred_label.asnumpy().astype('int32')
            label = label.asnumpy().astype('int32')

            check_label_shapes(label, pred_label)

            self.sum_metric += (pred_label.flat == label.flat).sum()
            self.num_inst += len(pred_label.flat)
```

Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues and bug reports. For non-technical issues and feature requests, feel free to present the information in what you believe is the best form.

For Q & A and discussion, please start a discussion thread at https://discuss.mxnet.io 

## Description
(Brief description of the problem in no more than 2 sentences.)

## Environment info (Required)

```
What to do:
1. Download the diagnosis script from https://raw.githubusercontent.com/apache/incubator-mxnet/master/tools/diagnose.py
2. Run the script using `python diagnose.py` and paste its output here.

```

Package used (Python/R/Scala/Julia):
(I'm using ...)

For Scala user, please provide:
1. Java version: (`java -version`)
2. Maven version: (`mvn -version`)
3. Scala runtime if applicable: (`scala -version`)

For R user, please provide R `sessionInfo()`:

## Build info (Required if built from source)

Compiler (gcc/clang/mingw/visual studio):

MXNet commit hash:
(Paste the output of `git rev-parse HEAD` here.)

Build config:
(Paste the content of config.mk, or the build command.)

## Error Message:
(Paste the complete error message, including stack trace.)

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1.
2.

## What have you tried to solve it?

1.
2.

**Comment by srochel** [comment]
Zack - please provide a complete (minimum) reproducer script, so fix can be validated.

**Comment by szha** [comment]
There was report about performance regression on volta 8gpu after switching to ndarray, so I'm reverting the ndarray acc change and reopening this issue for tracking.

**Comment by spidyDev** [comment]
@sandeep-krishnamurthy : Add Tag : Performance

**Comment by ThomasDelteil** [comment]
That seems pretty critical indeed. The previous PR wasn't correct, it was still doing the computation using asnumpy() because it was using asscalar() which behind the scenes is equivalent to asnumpy()[0].

**Comment by szha** [comment]
@ThomasDelteil as I clarified in your comment in the other PR, you probably misunderstood the logic in that PR. I think your comments really intend to say "non-blocking is faster" and I agree. I think the blocking logic for each metric.update is a design choice that's worth revisiting. Nonetheless, there will be more work to do in imperative execution before we can roll out a change that removes the blocking in metric.

**Comment by ThomasDelteil** [comment]
Yes that's correct. I see, ideally it would be nice to be able to opt-out of the blocking metrics. What is the rationale behind making the metrics blocking rather than non-blocking? You could always block on the `.get()`

**Comment by szha** [comment]
The blocking ensures that computation of next batch doesn't start before the computation of the current batch finishes. Without the blocking, in the current logic, imperative execution will continue by allocating new memory for new batch rather than reusing the memory.

**Comment by ThomasDelteil** [comment]
I see how that could be a problem, but I think that philosophically MXNet should let the users choose when to block. Typically I usually block after 100 batches or on every epoch. In a typical GPU with GBs of RAMs, the memory footprints of a few batches is negligible compared to the size of the network.

Also blocking or not blocking, I think we can get good gains by storing the accuracy on the right context. When running on GPU, I got a 100% improvement by storing the accuracy on the GPU.
Here is an example where I benchmarked the network trained in this tutorial:
https://github.com/zackchase/mxnet-the-straight-dope/blob/master/chapter04_convolutional-neural-networks/cnn-gluon.ipynb

That is the base implementation of the accuracy evaluation loop using mx.metric.accuracy
```
%%time
test_accuracy = evaluate_accuracy_blocking(train_data, net)
```
```
CPU times: user 1min 3s, sys: 2 s, total: 1min 5s
Wall time: 24.9 s
```

That is my implementation using only non-blocking operations
```
%%time
test_accuracy = evaluate_accuracy_non_blocking(train_data, net)
CPU times: user 50.5 s, sys: 1.24 s, total: 51.7 s
Wall time: 12 s
```

The same implementation but blocking on preds.wait_to_read() in the `.update()` function
```
%%time
test_accuracy = evaluate_accuracy_wait_to_read(train_data, net)
CPU times: user 56.6 s, sys: 1.18 s, total: 57.7 s
Wall time: 13.5 s
```
